### PR TITLE
feat: Dafny Plugins support

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR requires https://github.com/dafny-lang/dafny/pull/1817 to be merged so that it works properly

This PR features a new option "Dafny Plugins", which essentially is a list of comma-separated DLLs that Dafny can use in various places to augment its pipeline. It also supports arguments, as explained here:
https://github.com/dafny-lang/dafny/tree/master/Source/DafnyLanguageServer#plugins

Also, I added logging information about what launch arguments were used to start the DafnyLanguageServer.

### CODE REVIEW

**Comment by cpitclaudel** [comment]
I think this option should be a list instead of a comma-separated string

**Comment by MikaelMayer** [comment]
> I think this option should be a list instead of a comma-separated string

Unless you know something I'm not aware of, lists are not at all user friendly in VSCode. With a list, you get the interface:
image
which provides no clue on how to add this setting to settings.json (look how people would have to guess it's `dafny.compilerArgs` with this case)

If you make it a string, VSCode makes it easy to edit and use.

**Comment by cpitclaudel** [comment]
> Unless you know something I'm not aware of, lists are not at all user friendly in VSCode. 

https://code.visualstudio.com/updates/v1_37#_settings-editor-support-for-string-arrays

Maybe it's because you used an optional list instead of a list?

**Comment by camrein** [comment]
Thanks for linking that doc @cpitclaudel. I think the reason that we're not seeing this kind of UI interaction is that we do not specify the type of the items. Therefore, the `dafny.compilerArgs` option should look like the following:
```cs

        "dafny.compilerArgs": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "default": [
            "/verifyAllModules",
            "/compile:1",
            "/spillTargetCode:1"
          ],
          "description": "Optional array of strings as Dafny compilation arguments"
        }
```

**Comment by MikaelMayer** [comment]
> > Unless you know something I'm not aware of, lists are not at all user friendly in VSCode.
> 
> https://code.visualstudio.com/updates/v1_37#_settings-editor-support-for-string-arrays
> 
> Maybe it's because you used an optional list instead of a list?

Fantastic ! That's because the type of items was not specified. I'm going to do a separate PR so that it becomes the default.

image

Here is the PR. I'll update this PR accordingly.
https://github.com/dafny-lang/ide-vscode/pull/124

File: src/language/dafnyLanguageClient.ts
Comment (cpitclaudel) [inline src/language/dafnyLanguageClient.ts:60]: But on the Dafny side, don't we want to pass multiple copies of the flag instead of using a `,`?

**Review by cpitclaudel — CHANGES_REQUESTED** [review]
We need to wait for the Dafny side before we merge

File: src/language/dafnyLanguageClient.ts
Comment (MikaelMayer) [reply src/language/dafnyLanguageClient.ts]: Indeed, I forgot to push my changes for this VSCode plugin

**Review by camrein — CHANGES_REQUESTED** [review]
There are some linter warnings that should be resolved (see the results of the GitHub action). If you're using VSCode, install the ESLint extension.
You may also run the linter manually from CLI using `npm run lint`.

File: package.json
Comment (cpitclaudel) [inline package.json:164]: What happens if there's a comma in a plugin argument?  (Do we really need plugin arguments?)

File: src/language/dafnyLanguageClient.ts
Comment (cpitclaudel) [inline src/language/dafnyLanguageClient.ts:65]: This is dangerous.  I have double-checked the escaping, but it gets injected right into another command line, which is scary.

File: src/language/dafnyLanguageClient.ts
Comment (cpitclaudel) [inline src/language/dafnyLanguageClient.ts:69]: If there are command line arguments, they should be escaped before being concatenated, or better yet we should use an API that takes a list of arguments, not a string.

Comment (cpitclaudel) [reply src/language/dafnyLanguageClient.ts:69]: And in fact this is what we are doing, so this function should return a `string[]`, not a string (and in the caller you can use `...getDafnyPluginsArguments()`

File: src/language/dafnyLanguageClient.ts
Comment (cpitclaudel) [inline src/language/dafnyLanguageClient.ts:83]: This needs escaping

**Review by cpitclaudel — CHANGES_REQUESTED** [review]
Looks OK, but we shouldn't concatenate command line arguments

File: src/language/dafnyLanguageClient.ts
Comment (MikaelMayer) [reply src/language/dafnyLanguageClient.ts]: I removed this code, I now leave it entirely to the users to actually provide correct escape, which in most cases is not needed, and if needed, it matches the command-line escape that Dafny itself will unescape, not the plugin.

File: package.json
Comment (MikaelMayer) [reply package.json:164]: Only the first comma delimits the plugin from the argument list.
Yes we need plugin arguments, so that the same plugin can be configured through the class PluginConfiguration.
There are several use case for plugin arguments
* If the plugin provides several rewriters, to activate one of them only
* If rewriters have different options, such as "experimental" options, it makes it possible to activate them by changing a command-line flag.

**Comment by MikaelMayer** [comment]
I just push some changes but realized we need to simplify how the parsing of argument works at the Dafny level.

**Comment by MikaelMayer** [comment]
I just tested it again on one of the Dafny tests, it works like a charm.

File: src/language/dafnyLanguageClient.ts
Comment (camrein) [inline src/language/dafnyLanguageClient.ts:63]: Please always use curly braces.

You may add the following rule to `.eslintrc.json`:
```json
    "curly": "warn",
```

File: src/language/dafnyLanguageClient.ts
Comment (camrein) [inline src/language/dafnyLanguageClient.ts:82]: Please extract the "prettifying" of the arguments into a separate function for readability. The best would be to extract the whole part that's responsible to turn the `launchArguments` array into a string.

File: src/language/dafnyLanguageClient.ts
Comment (camrein) [inline src/language/dafnyLanguageClient.ts:61]: It should generally be avoided to use `for(... in ...)` to iterate over the indices of an array. For example, the order of the keys (which are the indices for arrays) is not consistent:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in

Threfore, prefer using a conventional for loop to iterate over the indices.

However, I think this whole for-loop can be neatly refactored into a map and filter call. Something like this:
```js
return plugins
  .filter(plugin => plugin != null && plugin !== '')
  .map((plugin, index) => `--dafny:plugins:${index}=${plugin}`);
```

**Review by camrein — CHANGES_REQUESTED** [review]

File: src/language/dafnyLanguageClient.ts
Comment (MikaelMayer) [reply src/language/dafnyLanguageClient.ts:61]: Well, I prefer comprehensions too. I also added parentheses around your expression because the semicolon is optional in JavaScript.

Comment (camrein) [reply src/language/dafnyLanguageClient.ts]: > Well, I prefer comprehensions too. I also added parentheses around your expression because the semicolon is optional in JavaScript.

Because the semicolons are optional in JavaScript/TypeScript, I configured eslint to enforce them. It's way to easy to forget about some implicit semicolon, especially when context switching between multiple languages.

File: src/language/dafnyLanguageClient.ts
Comment (camrein) [inline src/language/dafnyLanguageClient.ts:89]: Honestly, I don't like these random line breaks IntelliJ/Rider/WebStorm does, it should at least move the closing parenthesis to a new line too. Unfortunately, I haven't found an eslint rule to enforce this yet.

**Pull request merged**
